File: stepik-union/src/main/java/org/stepik/core/ui/FormListener.java
package org.stepik.core.ui;

import com.intellij.execution.ui.ConsoleViewContentType;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import javafx.application.Platform;
import javafx.stage.FileChooser;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.objects.submissions.Reply;
import org.stepik.api.objects.submissions.Submission;
import org.stepik.api.queries.submissions.StepikSubmissionsPostQuery;
import org.stepik.core.StepikProjectManager;
import org.stepik.core.StudyUtils;
import org.stepik.core.courseFormat.StepNode;
import org.stepik.core.courseFormat.StepType;
import org.stepik.core.courseFormat.StudyNode;
import org.stepik.core.stepik.StepikAuthManager;
import org.stepik.core.stepik.StepikAuthState;
import org.stepik.core.testFramework.toolWindow.StepikTestResultToolWindow;
import org.stepik.core.testFramework.toolWindow.StepikTestToolWindowUtils;
import org.stepik.plugin.actions.SendAction;
import org.stepik.plugin.actions.navigation.StudyNavigator;
import org.stepik.plugin.utils.NavigationUtils;
import org.w3c.dom.events.Event;
import org.w3c.dom.events.EventListener;
import org.w3c.dom.html.HTMLFormElement;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.List;
import java.util.stream.Collectors;

import static org.stepik.core.stepik.StepikAuthManager.authAndGetStepikApiClient;
import static org.stepik.core.ui.StepDescriptionUtils.getReply;
import static org.stepik.core.utils.ProjectFilesUtils.getOrCreateSrcDirectory;

class FormListener implements EventListener {
    static final String EVENT_TYPE_SUBMIT = "submit";
    private static final Logger logger = Logger.getInstance(FormListener.class);
    private final Project project;
    private final StudyBrowserWindow browser;

    FormListener(@NotNull Project project, @NotNull StudyBrowserWindow browser) {
        this.project = project;
        this.browser = browser;
    }

    @Nullable
    private static String getDataFromFile(@NotNull StepNode stepNode, @NotNull Project project) {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Open file");
        VirtualFile srcDirectory = getOrCreateSrcDirectory(project, stepNode, true);
        if (srcDirectory != null) {
            File initialDir = new File(srcDirectory.getPath());
            fileChooser.setInitialDirectory(initialDir);
        }
        File file = fileChooser.showOpenDialog(null);
        if (file != null) {
            try {
                List<String> lines = Files.readAllLines(file.toPath());
                return lines.stream().collect(Collectors.joining("\n"));
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return null;
    }

    static void getAttempt(@NotNull Project project, @NotNull StepNode node) {
        StepikApiClient stepikApiClient = authAndGetStepikApiClient(true);

        stepikApiClient.attempts()
                .post()
                .step(node.getId())
                .executeAsync()
                .whenComplete((attempts, e) -> {
                    if (attempts != null) {
                        node.cleanLastReply();
                        StepikProjectManager.updateSelection(project);
                    } else {
                        logger.warn(e);
                    }
                });
    }

    private static void sendStep(
            @NotNull Project project,
            @NotNull StepNode stepNode,
            @NotNull Elements elements,
            @NotNull StepType type,
            long attemptId,
            @Nullable String data) {
        StepikApiClient stepikApiClient = authAndGetStepikApiClient(true);

        StepikSubmissionsPostQuery query = stepikApiClient.submissions()
                .post()
                .attempt(attemptId);
        Reply reply = getReply(stepNode, type, elements, data);
        if (reply == null) {
            return;
        }

        query.reply(reply)
                .executeAsync()
                .whenComplete(((submissions, e) -> {
                    String lessonName;
                    StudyNode lesson = stepNode.getParent();
                    if (lesson != null) {
                        lessonName = lesson.getName();
                    } else {
                        lessonName = "";
                    }

                    final StepikTestResultToolWindow[] resultWindow = new StepikTestResultToolWindow[1];

                    String title = lessonName + " : " + stepNode.getName();
                    ApplicationManager.getApplication().invokeAndWait(() ->
                            resultWindow[0] = StepikTestToolWindowUtils.Companion
                                    .showTestResultsToolWindow(project, title));

                    resultWindow[0].clear();
                    resultWindow[0].println("Test method: send to Stepik", ConsoleViewContentType.NORMAL_OUTPUT);

                    if (submissions == null) {
                        printError(e, resultWindow[0]);
                        StepikProjectManager.updateSelection(project);
                        return;
                    }

                    if (submissions.isEmpty()) {
                        printError(e, resultWindow[0]);
                        return;
                    }

                    Submission submission = submissions.getFirst();
                    SendAction.INSTANCE.checkStepStatus(project,
                            stepikApiClient,
                            stepNode,
                            submission.getId(),
                            resultWindow[0]);
                }));
    }

    private static void printError(Throwable e, StepikTestResultToolWindow resultWindow) {
        resultWindow.println("Failed send step from browser", ConsoleViewContentType.ERROR_OUTPUT);
        resultWindow.println(e.getMessage(), ConsoleViewContentType.ERROR_OUTPUT);
        logger.warn("Failed send step from browser", e);
    }

    static void handle(
            @NotNull Project project,
            @NotNull StudyBrowserWindow browser,
            @NotNull HTMLFormElement form) {
        StudyNode root = StepikProjectManager.getProjectRoot(project);
        if (root == null) {
            return;
        }

        StudyNode node = StudyUtils.getStudyNode(root, form.getAction());
        if (!(node instanceof StepNode)) {
            return;
        }

        StepNode stepNode = (StepNode) node;
        Elements elements = new Elements(form.getElements());

        switch (elements.getAction()) {
            case "get_first_attempt":
            case "get_attempt":
                if (!elements.isLocked()) {
                    getAttempt(project, stepNode);
                }
                break;
            case "submit":
                String typeStr = elements.getType();
                StepType type = StepType.of(typeStr);
                boolean isFromFile = elements.isFromFile();
                String data = isFromFile ? getDataFromFile(stepNode, project) : null;
                long attemptId = elements.getAttemptId();
                sendStep(project, stepNode, elements, type, attemptId, data);
                break;
            case "save_reply":
                typeStr = elements.getType();
                type = StepType.of(typeStr);
                getReply(stepNode, type, elements, null);
                break;
            case "login":
                String email = elements.getInputValue("email");
                String password = elements.getInputValue("password");
                browser.showLoadAnimation();
                StepikAuthManager.authentication(email, password)
                        .whenComplete((state, throwable) -> {
                            if (state != StepikAuthState.AUTH) {
                                browser.callFunction("setErrorMessage", "Wrong email or password");
                            }
                            browser.hideLoadAnimation();
                        });
                break;
            case "next_step":
                StudyNode targetNode = StudyNavigator.nextLeaf(stepNode);

                if (targetNode == null) {
                    return;
                }

                Platform.runLater(() -> NavigationUtils.navigate(project, targetNode));
                break;
            default:
                browser.hideLoadAnimation();
        }
    }

    @Override
    public void handleEvent(Event event) {
        String domEventType = event.getType();
        if (EVENT_TYPE_SUBMIT.equals(domEventType)) {
            HTMLFormElement form = (HTMLFormElement) event.getTarget();
            handle(project, browser, form);
            event.preventDefault();
            event.stopPropagation();
        }
    }
}

File: stepik-union/src/main/kotlin/org/stepik/core/ui/FormListener.kt
package org.stepik.core.ui

import com.intellij.execution.ui.ConsoleViewContentType
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.project.Project
import javafx.application.Platform
import javafx.stage.FileChooser
import org.stepik.api.objects.ObjectsContainer
import org.stepik.core.StepikProjectManager
import org.stepik.core.StudyUtils
import org.stepik.core.actions.SendAction
import org.stepik.core.common.Loggable
import org.stepik.core.courseFormat.StepNode
import org.stepik.core.courseFormat.StepType
import org.stepik.core.stepik.StepikAuthManager
import org.stepik.core.stepik.StepikAuthManager.authAndGetStepikApiClient
import org.stepik.core.stepik.StepikAuthState
import org.stepik.core.testFramework.toolWindow.StepikTestResultToolWindow
import org.stepik.core.testFramework.toolWindow.StepikTestToolWindowUtils
import org.stepik.core.ui.StepDescriptionUtils.getReply
import org.stepik.core.utils.NavigationUtils
import org.stepik.core.utils.ProjectFilesUtils.getOrCreateSrcDirectory
import org.stepik.plugin.actions.navigation.StudyNavigator
import org.w3c.dom.events.Event
import org.w3c.dom.events.EventListener
import org.w3c.dom.html.HTMLFormElement
import java.io.File
import java.io.IOException
import java.nio.file.Files

internal class FormListener(private val project: Project, private val browser: StudyBrowserWindow) : EventListener {

    override fun handleEvent(event: Event) {
        val domEventType = event.type
        if (EVENT_TYPE_SUBMIT == domEventType) {
            val form = event.target as HTMLFormElement
            handle(project, browser, form)
            event.preventDefault()
            event.stopPropagation()
        }
    }

    companion object : Loggable {
        const val EVENT_TYPE_SUBMIT = "submit"

        private fun getDataFromFile(stepNode: StepNode, project: Project): String? {
            val fileChooser = FileChooser()
            fileChooser.title = "Open file"
            val srcDirectory = getOrCreateSrcDirectory(project, stepNode, true)
            if (srcDirectory != null) {
                val initialDir = File(srcDirectory.path)
                fileChooser.initialDirectory = initialDir
            }
            val file = fileChooser.showOpenDialog(null)
            if (file != null) {
                try {
                    val lines = Files.readAllLines(file.toPath())
                    return lines.joinToString { "\n" }
                } catch (e: IOException) {
                    logger.warn(e)
                }
            }
            return null
        }

        fun getAttempt(project: Project, node: StepNode) {
            val stepikApiClient = authAndGetStepikApiClient(true)

            stepikApiClient.attempts()
                    .post<ObjectsContainer<*>>()
                    .step(node.id)
                    .executeAsync()
                    .whenComplete { attempts, e ->
                        if (attempts != null) {
                            node.cleanLastReply()
                            StepikProjectManager.updateSelection(project)
                        } else {
                            logger.warn(e)
                        }
                    }
        }

        private fun sendStep(
                project: Project,
                stepNode: StepNode,
                elements: Elements,
                type: StepType,
                attemptId: Long,
                data: String?) {
            val stepikApiClient = authAndGetStepikApiClient(true)

            val query = stepikApiClient.submissions()
                    .post()
                    .attempt(attemptId)
            val reply = getReply(stepNode, type, elements, data) ?: return

            query.reply(reply)
                    .executeAsync()
                    .whenComplete { submissions, e ->
                        val lesson = stepNode.parent
                        val lessonName = lesson?.name ?: ""

                        val resultWindows = arrayOfNulls<StepikTestResultToolWindow>(1)

                        val title = "$lessonName : ${stepNode.name}"
                        ApplicationManager.getApplication().invokeAndWait {
                            resultWindows[0] = StepikTestToolWindowUtils
                                    .showTestResultsToolWindow(project, title)
                        }

                        val resultWindow = resultWindows[0]!!
                        resultWindow.clear()
                        resultWindow.println("Test method: send to Stepik")

                        if (submissions == null) {
                            printError(e, resultWindow)
                            StepikProjectManager.updateSelection(project)
                            return@whenComplete
                        }

                        if (submissions.isEmpty) {
                            printError(e, resultWindow)
                            return@whenComplete
                        }

                        val submission = submissions.first
                        SendAction.checkStepStatus(project,
                                stepikApiClient,
                                stepNode,
                                submission.id,
                                resultWindow)
                    }
        }

        private fun printError(e: Throwable, resultWindow: StepikTestResultToolWindow) {
            resultWindow.println("Failed send step from browser", ConsoleViewContentType.ERROR_OUTPUT)
            resultWindow.println(e.message ?: "", ConsoleViewContentType.ERROR_OUTPUT)
            logger.warn("Failed send step from browser", e)
        }

        fun handle(
                project: Project,
                browser: StudyBrowserWindow,
                form: HTMLFormElement) {
            val root = StepikProjectManager.getProjectRoot(project) ?: return

            val node = StudyUtils.getStudyNode(root, form.action) as? StepNode ?: return

            val elements = Elements(form.elements)

            when (elements.action) {
                "get_first_attempt", "get_attempt" -> if (!elements.isLocked) {
                    getAttempt(project, node)
                }
                "submit" -> {
                    val type = StepType.of(elements.type)
                    val isFromFile = elements.isFromFile
                    val data = if (isFromFile) getDataFromFile(node, project) else null
                    val attemptId = elements.attemptId
                    sendStep(project, node, elements, type, attemptId, data)
                }
                "save_reply" -> {
                    val type = StepType.of(elements.type)
                    getReply(node, type, elements, null)
                }
                "login" -> {
                    val email = elements.getInputValue("email")
                    val password = elements.getInputValue("password")
                    browser.showLoadAnimation()
                    StepikAuthManager.authentication(email, password)
                            .whenComplete { state, _ ->
                                if (state != StepikAuthState.AUTH) {
                                    browser.callFunction("setErrorMessage", "Wrong email or password")
                                }
                                browser.hideLoadAnimation()
                            }
                }
                "next_step" -> {
                    val targetNode = StudyNavigator.nextLeaf(node) ?: return

                    Platform.runLater { NavigationUtils.navigate(project, targetNode) }
                }
                else -> browser.hideLoadAnimation()
            }
        }
    }
}
